### DISCUSSION TITLE
Using NarrowLink + "ssh -j" as a massive jumphost system

### DISCUSSION THREAD

**Discussion by senpro-ingwersenk** [narrowlink/narrowlink#110] [Q&A] [answered]
Hello there!

We have ~20 customers who are equipped with a Raspberry Pi which we use to remotely monitor several aspects of their network. In order to manage them (they are NixOS based), we have to jump between VPNs.

My idea thus far is to utilize a miniature k3s cluster or Podman Compose in order to spawn a couple dozen of small containers that either link directly to the respective VPNs - or, see if I can use Narrowlink to simplify the deployment by a lot.

Here is the question:
Consider that I may be inclined to use a fly.io instance, so most management has to be done remotely. I would like for all these Pis to connect to that instance and:
- Not see each other or even be able to communicate with each other
- Not be able to connect to anything in the "network" but to be able to be connected to (so I can SSH into them, obvs)
How do I configure Narrowlink to do that?

Basically, I want to save myself time hopping VPNs. All of the Pis have a unique hostname that we share across clients; if Github was one of them, it'd be something like `gihuus-nd-mtag01`. So ultimatively what I would love to do in the end when this deployment is done is to simply use the `ProxyJump` feature to automatically jump to those Pis through NarrowLink.

Thank you very much and kind regards!

**Comment by SajjadPourali** [comment, marked as answer]
Hello

One of the objectives of Narrowlink is to prevent the need for a jumping host.

> Not see each other or even be able to communicate with each other

This is fundamental to Narrowlink’s architecture. Narrowlink agents work uni-directionally (one-way connection), which means agents don’t see or communicate with each other (out of the box).

> Not be able to connect to anything in the "network" but to be able to be connected to (so I can SSH into them, obvs) How do I configure Narrowlink to do that?

I didn’t understand this part well. The Narrowlink agent must connect to its server to establish its tunnel. However, the Narrowlink agent does not need to be installed on the end device. It can be installed on a device that can reach the end device and connect to it through the device that can see it.

For example the blue arrow in the bottom image demonstrate my explanation visually, both private server and agent are behind of the nat and narrowlink's agent is not installed on the private server.

<img  alt="narrowlink" src="https://narrowlink.com/assets/images/Diagram-3ea2363d1c9de62f21e08bac72686826.svg">


In a nutshell, Narrowlink agents don’t communicate with each other, and when you run them, you can connect to any host that they can reach. It is not necessary to install Narrowlink on every end-node device.

Narrowlink also integrates with SSH with a small configuration that automatically establishes a tunnel when you start a connection.

<img width="401" src="https://github.com/narrowlink/narrowlink/assets/20374762/99003d6b-e344-40b6-be2a-12499949fd30">

All of these machines are behind of a NAT network.

<img  src="https://github.com/narrowlink/narrowlink/assets/20374762/716e676d-c30f-491f-a83e-3be9eff454b9">


Please let me know if this helps, and don’t hesitate to ask any further questions.

**Reply by senpro-ingwersenk** [reply]
I had no idea there was a direct SSH integration - this makes this even more interesting for my use-case!

So technically, all the Raspberry Pis I want to connect to - or a device adjacent to it - would require a running agent that connects to the gateway (be it on fly.io or somewhere else). And my maschine (is "client" the right word here?) would need narrowlink installed as well in order to connect to the gateway, and thus to the Pi. Did I get that right?

I did notice that you used the verbatim `ssh` command - so, do you have anything specified in your `.ssh/config` to make this work, or are you just using local DNS resolution to make SSH utilize narrowlink for the connection?

Thank you for the very detailed answer! In parallel, I will also dig into the documentation more.

Kind regards!

**Reply by SajjadPourali** [reply]
> So technically, all the Raspberry Pis I want to connect to - or a device adjacent to it - would require a running agent that connects to the gateway (be it on fly.io or somewhere else). And my maschine (is "client" the right word here?) would need narrowlink installed as well in order to connect to the gateway, and thus to the Pi. Did I get that right?

True, however, one gateway can be used for all the agents that you have. The client and agent try to establish a peer-to-peer connection by default to reduce the load on the gateway. Therefore, practically, your SSH connection between the agent and client, when both are behind NAT, is direct (without passing through the gateway).

For SSH, as I was looking for a method that doesn’t affect the integrity of SSH itself and the OS, I implemented it as a proxy. So, you need to add the proxy form in .ssh/config as mentioned in [the documentation](https://narrowlink.com/docs/extended-tutorial/ssh-integration), or pass it as an argument, which I don’t suggest.

You are most welcome. Please let me know if you encounter any other questions.

**Reply by IngwiePhoenix** [reply]
Been reading the docs cover to back to get a better understanding of things and am currently in the process of setting it up on my private infra as a test case (I don't want to make my company buy a VM that is "just" used for testing). Sadly, Ring doesn't support RISC-V, which is a good bit of my infra. Oh well.

While poking around with the client on Windows, I noticed that nowhere in it's help pages does it mention the default path for the configuration. I didn't set one on purpose to see what would happen; my collegues aren't home in the shell as much as me, so this would help if:
```
PS C:\tools\bin> .\narrowlink.exe list
2024-06-18T06:59:39.377631Z ERROR narrowlink: Error: Config Not Found
```
...would also mention the default location. I.e.: `Config not found ($CONFIG_PATH)`. I'll see if I can make a PR for that, actually - seems simple enough.

Aside from that, I will try to figure everything else out from the docs; will have to try my luck with policies first because those are one of the most important bits (restricting to just port 22 since this is all we need/use).

I am also looking at creating a small `fly.toml`. Their free-tier VMs/containers should be quite sufficient as a gateway; if I have something working, I will send it your way as an example deployment.

Thank you very much for your insights! Much appreciated!

**Reply by senpro-ingwersenk** [reply]
Oops, forgot to switch to my work account. Same difference though; thanks for the help!

**Reply by SajjadPourali** [reply]
> Been reading the docs cover to back to get a better understanding of things and am currently in the process of setting it up on my private infra as a test case (I don't want to make my company buy a VM that is "just" used for testing). Sadly, Ring doesn't support RISC-V, which is a good bit of my infra. Oh well.
> 
> While poking around with the client on Windows, I noticed that nowhere in it's help pages does it mention the default path for the configuration. I didn't set one on purpose to see what would happen; my collegues aren't home in the shell as much as me, so this would help if:
> 
> ```
> PS C:\tools\bin> .\narrowlink.exe list
> 2024-06-18T06:59:39.377631Z ERROR narrowlink: Error: Config Not Found
> ```
> 
> ...would also mention the default location. I.e.: `Config not found ($CONFIG_PATH)`. I'll see if I can make a PR for that, actually - seems simple enough.
> 
> Aside from that, I will try to figure everything else out from the docs; will have to try my luck with policies first because those are one of the most important bits (restricting to just port 22 since this is all we need/use).
> 
> I am also looking at creating a small `fly.toml`. Their free-tier VMs/containers should be quite sufficient as a gateway; if I have something working, I will send it your way as an example deployment.
> 
> Thank you very much for your insights! Much appreciated!

A tiny server with ~512MB RAM is also capable of playing the role of the gateway. However, you can drop an email to try@narrowlink.com to get tokens for the client and agent for your test: https://narrowlink.com/docs/intro#user-content-fnref-4 .

The configuration path is mentioned in [the documentation](https://narrowlink.com/docs/client#default-configuration-paths). However, I have just opened a feature request for the environment variables: https://github.com/narrowlink/narrowlink/issues/112 .

Narrowlink will also support other architectures soon. However, there are some considerations about this that will be explained here: https://github.com/narrowlink/narrowlink/issues/89 .

Please let me know if you encounter any other questions.
